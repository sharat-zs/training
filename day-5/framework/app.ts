import Koa, { Next } from 'koa'
import Router, { RouterContext } from 'koa-router'
import http, { Server } from 'http'
import bodyParser from 'koa-bodyparser'
import { Context, HttpMethod, HTTPStatus, IApp, RouteHandler } from './types'

export default class App implements IApp {
    private readonly koa = new Koa()

    private readonly router = new Router()

    constructor() {
        this.koa.use(bodyParser())
        this.koa.use(this.router.routes())
        this.koa.use(this.router.allowedMethods())
    }

    get(path: string, handler: RouteHandler) {
        this.registerRoute(HttpMethod.Get, path, handler)
        return this
    }

    put(path: string, handler: RouteHandler) {
        this.registerRoute(HttpMethod.Put, path, handler)
        return this
    }

    post(path: string, handler: RouteHandler) {
        this.registerRoute(HttpMethod.Post, path, handler)
        return this
    }

    delete(path: string, handler: RouteHandler) {
        this.registerRoute(HttpMethod.Delete, path, handler)
        return this
    }

    start(port: number = 8000) {
        return new Promise<Server>((resolve) => {
            const httpServer = this.koa.listen(
                http.createServer(this.koa.callback()).listen(port),
                () => {
                    console.info(`http server started on port ${port}`)
                    resolve(httpServer)
                }
            )
        })
    }

    private registerRoute(
        method: HttpMethod,
        path: string,
        handler: RouteHandler
    ) {
        this.router.register(path, [method], createRouteMiddleware(handler))
    }
}

function createRouteMiddleware(handler: RouteHandler) {
    return async (ctx: RouterContext, next: Next) => {
        try {
            const res = await handler(transformContext(ctx))
            let status = HTTPStatus.OK
            let body = res
            const { method } = ctx.request
            if (method === 'POST') status = HTTPStatus.CREATED
            if (method === 'DELETE') status = HTTPStatus.NO_CONTENT
            body = { data: res, status }
            if (!res) body = undefined
            ctx.response.body = body
            ctx.response.status = status
        } catch (err) {
            console.error(err)
            ctx.response.body = {
                message: err.message ?? 'Internal Server Error',
                code: err.code ?? 'UNKOWN_ERROR'
            }
            ctx.response.status = err.status ?? 500
        }

        return next()
    }
}

function transformContext(ctx: RouterContext): Context {
    return {
        headers: ctx.request.headers,
        query: ctx.request.query,
        body: ctx.request.body,
        params: ctx.params,
        path: ctx.path
    }
}
