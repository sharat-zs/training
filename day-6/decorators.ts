import 'reflect-metadata'

/* eslint-disable no-param-reassign */
export class Person {
    @format('hello, %s')
    dummy: string

    constructor(readonly name: string, readonly age: number) {
        this.greet = this.greet.bind(this)
        this.dummy = 'dummmy'
    }

    @log()
    greet() {
        const formatString = getFormat(this, 'dummy')
        return formatString.replace('%s', this.dummy)
    }
}

function log() {
    return function (
        _target: any,
        _propertyName: any,
        descriptor: TypedPropertyDescriptor<() => string>
    ) {
        const method = descriptor.value
        if (!method) throw new Error('Should never happen, method not defined')
        descriptor.value = function () {
            console.log('method called')
            const res = method.call(this)
            return res
        }
    }
}

function format(formatString: string) {
    const formatMetadataKey = 'Format'
    return Reflect.metadata(formatMetadataKey, formatString)
}

function getFormat(target: any, propertyKey: string) {
    const formatMetadataKey = 'Format'
    return Reflect.getMetadata(formatMetadataKey, target, propertyKey)
}

const person = new Person('harry', 18)
console.log(person.greet())
