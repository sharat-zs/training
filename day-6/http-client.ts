import axios from 'axios'

export const URL = 'https://jsonplaceholder.typicode.com/todos/1'
export async function fetchData(url: string) {
    const res = await axios.get(url)
    const data = res.data
    const status = res.status
    return { data, status }
}

// fetchData(URL).then((res) => console.log(res))
