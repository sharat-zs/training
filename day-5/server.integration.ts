import { Server } from 'http'
import supertest from 'supertest'
import axios from 'axios'
import AxiosMockAdapter from 'axios-mock-adapter'
import app from './server'
import { URL } from '../day-6/http-client'

// jest.mock('axios')
// const mockedAxios = axios as jest.Mocked<typeof axios>
const mockedAxios = new AxiosMockAdapter(axios)
mockedAxios.onGet(URL).reply(200, {
    users: [{ id: 1, name: 'John Smith' }]
})

describe('integration tests', () => {
    let server: Server
    let agent: supertest.SuperTest<supertest.Test>
    expect.hasAssertions()

    beforeAll(async () => {
        server = await app
        agent = supertest(server)
    })

    afterAll(() => {
        server.unref()
        server.close()
    })

    it('should send 200 response to "/hello"', async () => {
        const res = agent.get('/hello')
        expect((await res).status).toBe(200)
        expect((await res).body).toMatchSnapshot()
    })

    it('should always throw error', async () => {
        const res = agent.get('/error')
        expect((await res).status).toBe(400)
        expect((await res).body).toMatchSnapshot()
    })

    it('should return 200 on outbound', async () => {
        const res = agent.get('/outgoing')
        expect((await res).status).toBe(200)
        expect((await res).body.data).toMatchSnapshot()
    })
})
