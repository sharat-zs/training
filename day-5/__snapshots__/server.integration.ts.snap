// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests should always throw error 1`] = `
Object {
  "code": "KNOWN_ERROR",
  "message": "Known error, won't fix",
}
`;

exports[`integration tests should return 200 on outbound 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": 1,
        "name": "John Smith",
      },
    ],
  },
  "status": 200,
}
`;

exports[`integration tests should send 200 response to "/hello" 1`] = `
Object {
  "data": "hello",
  "status": 200,
}
`;
